src/l1/blob_fetcher.rs

```diff
@@ -0,0 +1,286 @@
+use std::sync::{
+    atomic::{AtomicU64, Ordering},
+    Arc,
+};
+
+use bytes::Bytes;
+use ethers::types::{Block, Transaction};
+use eyre::Result;
+use serde::Deserialize;
+use serde_json::Value;
+
+use super::decode_blob_data;
+use crate::config::Config;
+
+/// The transaction type used to identify transactions that carry blobs
+/// according to EIP 4844.
+const BLOB_CARRYING_TRANSACTION_TYPE: u64 = 3;
+
+/// The data contained in a batcher transaction.
+/// The actual source of this data can be either calldata or blobs.
+pub type BatcherTransactionData = Bytes;
+
+/// The blob fetcher is responsible for fetching blob data from the L1 beacon chain,
+/// along with relevant parsing and validation.
+///
+/// Consensus layer info required for deriving the slot at which a specific blob was
+/// included in the beacon chain is fetched on the first call to [`Self::get_slot_from_time`]
+/// and cached for all subsequent calls.
+pub struct BlobFetcher {
+    config: Arc<Config>,
+    client: reqwest::Client,
+    genesis_timestamp: AtomicU64,
+    seconds_per_slot: AtomicU64,
+}
+
+/// A beacon chain blob sidecar object.
+/// KZG commitment and proof fields are not used in the current implementation.
+#[derive(Debug, Deserialize)]
+pub struct BlobSidecar {
+    /// Blob index (transactions can have more than one blob)
+    #[serde(deserialize_with = "deserialize_string_to_u64")]
+    pub index: u64,
+    /// Blob data (not decoded)
+    #[serde(deserialize_with = "deserialize_blob_bytes")]
+    pub blob: Vec<u8>,
+}
+
+impl BlobFetcher {
+    /// Create a new blob fetcher with the given config.
+    pub fn new(config: Arc<Config>) -> Self {
+        Self {
+            config,
+            client: reqwest::Client::new(),
+            genesis_timestamp: AtomicU64::new(0),
+            seconds_per_slot: AtomicU64::new(0),
+        }
+    }
+
+    /// Given a block, return a list of `BatcherTransactionData` containing either the
+    /// calldata or the decoded blob data for each batcher transaction in the block.
+    pub async fn get_batcher_transactions(
+        &self,
+        block: &Block<Transaction>,
+    ) -> Result<Vec<BatcherTransactionData>> {
+        let mut batcher_transactions_data = Vec::new();
+        let mut indexed_blobs = Vec::new();
+        let mut blob_index = 0;
+
+        for tx in block.transactions.iter() {
+            let tx_blob_hashes: Vec<String> = tx
+                .other
+                .get_deserialized("blobVersionedHashes")
+                .unwrap_or(Ok(Vec::new()))
+                .unwrap_or_default();
+
+            if !self.is_valid_batcher_transaction(tx) {
+                blob_index += tx_blob_hashes.len();
+                continue;
+            }
+
+            let tx_type = tx.transaction_type.map(|t| t.as_u64()).unwrap_or(0);
+            if tx_type != BLOB_CARRYING_TRANSACTION_TYPE {
+                batcher_transactions_data.push(tx.input.0.clone());
```

Comment:
Since this module handles fetching both blob batches and non blob batches we might want to consider renaming or potentially separating out the functionality into two areas. It took me a bit to figure out where regular batches were decoded as the naming threw me off.

Reply:
You're right